### PR TITLE
New Serverless Pattern - Amazon Cognito user management using Amazon API Gateway and AWS Lambda

### PR DESCRIPTION
**Pull request by biswanathmukherjee**
*Issue #, if available:*

*Description of changes:*

Amazon Cognito user management using Amazon API Gateway and AWS Lambda
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: apigw-lambda-cognito-sam-java/README.md [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+3. **Amazon Cognito**: Depending on the operation invoked, user is created, confirmed or logged in the Amazon Cognito user pool.
+
+4. **Response**: The Lambda function processes the Amazon Cognito API response and sends it back to the user via the API Gateway.
+
+## Testing
+
+1. You can use [curl](https://curl.se/) or any other tool of your choice to send a HTTP POST create user request to the API. Make sure to replace `Endpoint` with the corresponding values from your `sam deploy --guided` output. Also replace all other values. The `email` should be a valid email id.
+
+   ```bash
+   curl -d '{"firstName": "{YourFirstName}", "lastName": "{YourLastName}", "email": "{your-email@mail.com}", "password": "{Y0urP#ssword}"}' -H 'Content-Type: application/json' {CreateUserAPIGatewayEndpoin}
```
Comment (bfreiberg) [inline apigw-lambda-cognito-sam-java/README.md:80]: Please leave the password value empty and rather suggest to use a secure value

File: apigw-lambda-cognito-sam-java/README.md [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+   ```
+
+   The API returns a response as below: 
+
+   ```json
+   {"isSuccessful":true,"statusCode":200,"cognitoUserId":"84c8xxxx-xxxx-xxxx-xxx-xxxxx9846de1","isConfirmed":false}
+   ```
+
+2. You should get an email with a verification code in the provided email id. Use the varification code to invoke the confirm user POST API:
```
Comment (bfreiberg) [inline apigw-lambda-cognito-sam-java/README.md:89]:
```suggestion
2. You should receive an email with a verification code at your provided email. Use the verification code to invoke the confirm user POST API:
```

File: apigw-lambda-cognito-sam-java/README.md [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+   The API returns a response as below: 
+
+   ```json
+   {"isSuccessful":true,"statusCode":200}
+   ```
+
+3. Now, you can call the login POST API. This API will return the token:
+You should get an email with a verification code in the provided email id. Use the varification code to invoke the confirm user POST API:
+   ```bash
+   curl -d '{"email": "{your-email@mail.com}", "password": "{Y0urP#ssword}"}' -H 'Content-Type: application/json' {LoginUserAPIGatewayEndpoint}
```
Comment (bfreiberg) [inline apigw-lambda-cognito-sam-java/README.md:103]: See above

File: apigw-lambda-cognito-sam-java/dependency-reduced-pom.xml [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+  <modelVersion>4.0.0</modelVersion>
+  <groupId>com.example</groupId>
+  <artifactId>apigw-lambda-cognito</artifactId>
+  <name>Cognito user management</name>
+  <version>1.0-SNAPSHOT</version>
+  <build>
+    <plugins>
+      <plugin>
+        <artifactId>maven-surefire-plugin</artifactId>
+        <version>3.0.0-M5</version>
```
Comment (bfreiberg) [inline apigw-lambda-cognito-sam-java/dependency-reduced-pom.xml:12]: any reason not to use the latest version?

File: apigw-lambda-cognito-sam-java/dependency-reduced-pom.xml [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+      </plugin>
+      <plugin>
+        <artifactId>maven-shade-plugin</artifactId>
+        <version>3.2.4</version>
```
Comment (bfreiberg) [inline apigw-lambda-cognito-sam-java/dependency-reduced-pom.xml:16]: any reason not to use the latest version?

File: apigw-lambda-cognito-sam-java/dependency-reduced-pom.xml [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+        <configuration />
+      </plugin>
+    </plugins>
+  </build>
+  <dependencyManagement>
+    <dependencies>
+      <dependency>
+        <groupId>software.amazon.awssdk</groupId>
+        <artifactId>bom</artifactId>
+        <version>2.21.19</version>
```
Comment (bfreiberg) [inline apigw-lambda-cognito-sam-java/dependency-reduced-pom.xml:34]: any reason not to use the latest version?

File: apigw-lambda-cognito-sam-java/pom.xml [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+        <maven.compiler.target>21</maven.compiler.target>
+        <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
+    </properties>
+
+    <dependencyManagement>
+        <dependencies>
+            <dependency>
+                <groupId>software.amazon.awssdk</groupId>
+                <artifactId>bom</artifactId>
+                <version>2.25.38</version>
```
Comment (bfreiberg) [inline apigw-lambda-cognito-sam-java/pom.xml:22]: see above regarding versioning

File: apigw-lambda-cognito-sam-java/pom.xml [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+            <artifactId>gson</artifactId>
+            <version>2.10.1</version>
+        </dependency>
+    </dependencies>
+    <build>
+        <plugins>
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-surefire-plugin</artifactId>
+                <version>3.0.0-M5</version>
```
Comment (bfreiberg) [inline apigw-lambda-cognito-sam-java/pom.xml:74]: see above

File: apigw-lambda-cognito-sam-java/pom.xml [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+            </plugin>
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-shade-plugin</artifactId>
+                <version>3.2.4</version>
```
Comment (bfreiberg) [inline apigw-lambda-cognito-sam-java/pom.xml:79]: see above

File: apigw-lambda-cognito-sam-java/template.yaml [hunk 10]
```diff
@@ -0,0 +1,138 @@
+AWSTemplateFormatVersion: 2010-09-09
+Transform: AWS::Serverless-2016-10-31
+Description: This template exposes API to create and confirm user in Amazon Cognito user pool
+
+Globals:
+  Function:
+    Runtime: java21
+    MemorySize: 3008
+    Timeout: 120
```
Comment (bfreiberg) [inline apigw-lambda-cognito-sam-java/template.yaml:9]: Is that necessary with an API Gateway timeout of 29 seconds?

File: apigw-lambda-cognito-sam-java/README.md (hunk 1 above)
Comment (biswanathmukherjee) [reply apigw-lambda-cognito-sam-java/README.md]: Fixed

File: apigw-lambda-cognito-sam-java/README.md (hunk 2 above)
Comment (biswanathmukherjee) [reply apigw-lambda-cognito-sam-java/README.md]: Fixed.

File: apigw-lambda-cognito-sam-java/README.md (hunk 3 above)
Comment (biswanathmukherjee) [reply apigw-lambda-cognito-sam-java/README.md]: Fixed.

File: apigw-lambda-cognito-sam-java/dependency-reduced-pom.xml (hunk 4 above)
Comment (biswanathmukherjee) [reply apigw-lambda-cognito-sam-java/dependency-reduced-pom.xml:12]: Fixed.

File: apigw-lambda-cognito-sam-java/dependency-reduced-pom.xml (hunk 5 above)
Comment (biswanathmukherjee) [reply apigw-lambda-cognito-sam-java/dependency-reduced-pom.xml:16]: Fixed.

File: apigw-lambda-cognito-sam-java/dependency-reduced-pom.xml (hunk 6 above)
Comment (biswanathmukherjee) [reply apigw-lambda-cognito-sam-java/dependency-reduced-pom.xml:34]: Fixed.

File: apigw-lambda-cognito-sam-java/pom.xml (hunk 7 above)
Comment (biswanathmukherjee) [reply apigw-lambda-cognito-sam-java/pom.xml]: Fixed.

File: apigw-lambda-cognito-sam-java/pom.xml (hunk 8 above)
Comment (biswanathmukherjee) [reply apigw-lambda-cognito-sam-java/pom.xml]: Fixed.

File: apigw-lambda-cognito-sam-java/pom.xml (hunk 9 above)
Comment (biswanathmukherjee) [reply apigw-lambda-cognito-sam-java/pom.xml]: Fixed.

File: apigw-lambda-cognito-sam-java/template.yaml (hunk 10 above)
Comment (biswanathmukherjee) [reply apigw-lambda-cognito-sam-java/template.yaml]: Fixed.

File: apigw-lambda-cognito-sam-java/dependency-reduced-pom.xml (hunk 4 above)
Comment (bfreiberg) [reply apigw-lambda-cognito-sam-java/dependency-reduced-pom.xml:12]: According to https://mvnrepository.com/artifact/org.apache.maven.plugins/maven-surefire-plugin the latest version is `3.3.1`?

File: apigw-lambda-cognito-sam-java/dependency-reduced-pom.xml (hunk 5 above)
Comment (bfreiberg) [reply apigw-lambda-cognito-sam-java/dependency-reduced-pom.xml:16]: Similar to above, https://mvnrepository.com/artifact/org.apache.maven.plugins/maven-shade-plugin has `3.6.0` as the latest

File: apigw-lambda-cognito-sam-java/dependency-reduced-pom.xml (hunk 6 above)
Comment (bfreiberg) [reply apigw-lambda-cognito-sam-java/dependency-reduced-pom.xml:34]: See https://mvnrepository.com/artifact/software.amazon.awssdk/bom

Comment (biswanathmukherjee) [reply apigw-lambda-cognito-sam-java/dependency-reduced-pom.xml]: Fixed the pom.xml file and removed dependency-reduced-pom.xml as that is not required to be checked in.

File: apigw-lambda-cognito-sam-java/dependency-reduced-pom.xml (hunk 4 above)
Comment (biswanathmukherjee) [reply apigw-lambda-cognito-sam-java/dependency-reduced-pom.xml]: Fixed the pom.xml file and removed dependency-reduced-pom.xml as that is not required to be checked in.

**Comment by bfreiberg** [comment]
Looks good, thanks for your contribution. Your PR will be merged soon

**Pull request merged**